File: android/src/main/kotlin/io/sentient/android/settings/secrets/SecretsScreen.kt
// ---------------------------------------------------------------------------
// SecretsScreen — the Secrets (admin) settings page (Admin group). Per-provider
// masked key rows (OpenRouter / Ollama Cloud / Custom + base URL) showing presence
// only, "Set active" per row, and — after any save — an inline "restart to pick up
// the new key" notice with an "Apply now" button that fires the bare apply. State +
// callbacks come from [SecretsViewModel]; the key drafts live in SecretRow, never
// here. testTags: settings-secrets-screen/-back, settings-secrets-apply, plus the
// per-provider row tags from SecretRow.
// ---------------------------------------------------------------------------
package io.sentient.android.settings.secrets

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.safeDrawingPadding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Button
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.tooling.preview.Preview
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import io.sentient.android.settings.components.SettingsCard
import io.sentient.android.settings.components.SettingsTopBar
import io.sentient.android.theme.LocalTokens
import io.sentient.android.theme.SentientTheme
import io.sentient.mobiledata.usecase.settings.ApplyState
import io.sentient.mobilesdk.design.Colors
import io.sentient.mobilesdk.settings.LlmSecretsStatus
import io.sentient.mobilesdk.settings.SecretsStatus

private const val TITLE = "Secrets"

@Composable
fun SecretsScreen(
    vm: SecretsViewModel,
    onBack: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val state by vm.state.collectAsStateWithLifecycle()
    SecretsContent(
        state = state,
        onBack = onBack,
        onStartEdit = vm::startEdit,
        onCancelEdit = vm::cancelEdit,
        onSaveKey = vm::saveKey,
        onSaveBaseUrl = vm::saveBaseUrl,
        onSetActive = vm::setActive,
        onApplyNow = vm::applyNow,
        modifier = modifier,
    )
}

@Composable
private fun SecretsContent(
    state: SecretsUiState,
    onBack: () -> Unit,
    onStartEdit: (String, SecretField) -> Unit,
    onCancelEdit: () -> Unit,
    onSaveKey: (String, String) -> Unit,
    onSaveBaseUrl: (String) -> Unit,
    onSetActive: (String) -> Unit,
    onApplyNow: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val tokens = LocalTokens.current
    Column(
        modifier = modifier.fillMaxSize().safeDrawingPadding().testTag("settings-secrets-screen"),
    ) {
        SettingsTopBar(title = TITLE, onBack = onBack, backTestTag = "settings-secrets-back")
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .verticalScroll(rememberScrollState())
                .padding(horizontal = tokens.space.lg, vertical = tokens.space.md),
            verticalArrangement = Arrangement.spacedBy(tokens.space.md),
        ) {
            val status = state.status
            if (status == null) {
                Text("Loading…", color = Color(Colors.ink3), fontSize = tokens.type.sm)
            } else {
                ProviderCard(status, state.editing, onStartEdit, onCancelEdit, onSaveKey, onSaveBaseUrl, onSetActive)
                RestartNotice(state.restartNeeded, state.applying, onApplyNow)
            }
            if (state.errorMessage != null) {
                Text(state.errorMessage, color = Color(Colors.stop), fontSize = tokens.type.sm)
            }
        }
    }
}

@Composable
private fun ProviderCard(
    status: SecretsStatus,
    editing: EditingTarget?,
    onStartEdit: (String, SecretField) -> Unit,
    onCancelEdit: () -> Unit,
    onSaveKey: (String, String) -> Unit,
    onSaveBaseUrl: (String) -> Unit,
    onSetActive: (String) -> Unit,
) {
    val llm: LlmSecretsStatus = status.llm
    SettingsCard(title = "Active keys", subtitle = "Encrypted at rest. Shared by the household gateway.") {
        SecretRow(
            label = "OpenRouter",
            hasKey = llm.openrouter.hasKey,
            isActive = llm.active == SecretProviders.OPENROUTER,
            editing = editing.matches(SecretProviders.OPENROUTER, SecretField.KEY),
            onSetActive = { onSetActive(SecretProviders.OPENROUTER) },
            onStartEdit = { onStartEdit(SecretProviders.OPENROUTER, SecretField.KEY) },
            onCancel = onCancelEdit,
            onSave = { onSaveKey(SecretProviders.OPENROUTER, it) },
            testTagBase = "settings-secrets-openrouter",
        )
        SecretRow(
            label = "Ollama Cloud",
            hasKey = llm.ollamaCloud.hasKey,
            isActive = llm.active == SecretProviders.OLLAMA_CLOUD,
            editing = editing.matches(SecretProviders.OLLAMA_CLOUD, SecretField.KEY),
            onSetActive = { onSetActive(SecretProviders.OLLAMA_CLOUD) },
            onStartEdit = { onStartEdit(SecretProviders.OLLAMA_CLOUD, SecretField.KEY) },
            onCancel = onCancelEdit,
            onSave = { onSaveKey(SecretProviders.OLLAMA_CLOUD, it) },
            testTagBase = "settings-secrets-ollama",
        )
        SecretRow(
            label = "Custom",
            hasKey = llm.custom.hasKey,
            isActive = llm.active == SecretProviders.CUSTOM,
            editing = editing.matches(SecretProviders.CUSTOM, SecretField.KEY),
            onSetActive = { onSetActive(SecretProviders.CUSTOM) },
            onStartEdit = { onStartEdit(SecretProviders.CUSTOM, SecretField.KEY) },
            onCancel = onCancelEdit,
            onSave = { onSaveKey(SecretProviders.CUSTOM, it) },
            testTagBase = "settings-secrets-custom",
        )
        BaseUrlRow(
            hasBaseUrl = llm.custom.hasBaseUrl,
            editing = editing.matches(SecretProviders.CUSTOM, SecretField.BASE_URL),
            onStartEdit = { onStartEdit(SecretProviders.CUSTOM, SecretField.BASE_URL) },
            onCancel = onCancelEdit,
            onSave = onSaveBaseUrl,
            testTagBase = "settings-secrets-custom",
        )
    }
}

@Composable
private fun RestartNotice(restartNeeded: Boolean, applying: ApplyState?, onApplyNow: () -> Unit) {
    if (!restartNeeded && applying == null) return
    val tokens = LocalTokens.current
    val message = when (applying) {
        ApplyState.Restarting -> "Applying — assistant restarting…"
        ApplyState.AlreadyApplying -> "Already applying — another change is in progress."
        is ApplyState.Failed -> "Apply failed. Try again."
        is ApplyState.Ready, null -> "Restart the assistant to pick up the new key."
        ApplyState.Idle, ApplyState.Saving -> "Restart the assistant to pick up the new key."
    }
    Column(
        modifier = Modifier.fillMaxWidth().testTag("settings-secrets-restart-notice"),
        verticalArrangement = Arrangement.spacedBy(tokens.space.sm),
    ) {
        Text(message, color = Color(Colors.warn), fontSize = tokens.type.sm)
        if (applying !is ApplyState.Restarting) {
            Button(onClick = onApplyNow, modifier = Modifier.testTag("settings-secrets-apply")) {
                Text(if (applying is ApplyState.Failed) "Retry" else "Apply now")
            }
        }
    }
}

private fun EditingTarget?.matches(provider: String, field: SecretField): Boolean =
    this != null && this.provider == provider && this.field == field

@Preview
@Composable
private fun SecretsScreenPreview() {
    SentientTheme {
        SecretsContent(
            state = SecretsUiState(
                status = SecretsStatus(
                    llm = LlmSecretsStatus(
                        active = SecretProviders.OPENROUTER,
                        openrouter = io.sentient.mobilesdk.settings.LlmProviderStatus(hasKey = true),
                    ),
                ),
                restartNeeded = true,
            ),
            onBack = {}, onStartEdit = { _, _ -> }, onCancelEdit = {}, onSaveKey = { _, _ -> },
            onSaveBaseUrl = {}, onSetActive = {}, onApplyNow = {},
        )
    }
}

File: ios/App/Settings/Secrets/SecretsScreen.swift
import SwiftUI
import MobileData

struct SecretsScreen: View {
    @State private var vm: SecretsViewModel

    init(settings: SettingsComponent, onBack: @escaping () -> Void) {
        _vm = State(initialValue: SecretsViewModel(admin: settings.admin, applyProfileChange: settings.applyProfileChange))
    }

    var body: some View {
        SecretsBody(
            status: vm.status,
            isError: vm.isError,
            isNotAdmin: vm.isNotAdmin,
            mutationError: vm.mutationError,
            editing: vm.editing,
            isSaving: vm.isSavingKey,
            apply: vm.apply,
            onRetryLoad: { Task { await vm.load() } },
            onSetActive: { provider in Task { await vm.setActive(provider) } },
            onStartEditKey: vm.startEditKey,
            onStartEditBaseUrl: vm.startEditBaseUrl,
            onCancelEdit: vm.cancelEdit,
            onSaveKey: { provider, value in Task { await vm.saveKey(provider, value: value) } },
            onSaveBaseUrl: { url in Task { await vm.saveBaseUrl(url) } },
            onApplyNow: { Task { await vm.applyNow() } },
            onDismissNotice: vm.dismissNotice
        )
        .task { await vm.load() }
    }
}

private struct SecretsBody: View {
    let status: SecretsStatus?
    let isError: Bool
    let isNotAdmin: Bool
    let mutationError: String?
    let editing: SecretsViewModel.EditTarget
    let isSaving: Bool
    let apply: SecretsViewModel.ApplyPhase
    let onRetryLoad: () -> Void
    let onSetActive: (SecretsViewModel.Provider) -> Void
    let onStartEditKey: (SecretsViewModel.Provider) -> Void
    let onStartEditBaseUrl: () -> Void
    let onCancelEdit: () -> Void
    let onSaveKey: (SecretsViewModel.Provider, String) -> Void
    let onSaveBaseUrl: (String) -> Void
    let onApplyNow: () -> Void
    let onDismissNotice: () -> Void

    var body: some View {
        SettingsPageScaffold(title: "Secrets", screenId: "settings-secrets-screen") {
            if let mutationError {
                AsyncNotice(kind: .error, title: "Secret change failed", detail: mutationError)
            }
            content
        }
    }

    @ViewBuilder private var content: some View {
        if isNotAdmin {
            AsyncNotice(
                kind: .warning,
                title: "Admin access required",
                detail: "Your access may have changed. Secret controls are no longer available.",
                retry: onRetryLoad
            )
        } else if let status {
            providerSummary(status)
            DesignCard(
                title: "Provider keys",
                detail: "Key presence only — not a check that credentials work. Saved values are never displayed.",
                headerStyle: .quiet,
                bodyStyle: .rows
            ) {
                keyRow(.openrouter, label: "OpenRouter", status: status, hasKey: status.llm.openrouter.hasKey)
                DesignDivider()
                keyRow(.ollamaCloud, label: "Ollama Cloud", status: status, hasKey: status.llm.ollamaCloud.hasKey)
            }
            DesignCard(
                title: "Custom provider",
                detail: "The Custom key and base URL belong together. Save each value here, then select Custom to save your provider selection.",
                headerStyle: .quiet,
                bodyStyle: .rows
            ) {
                keyRow(.custom, label: "Custom", status: status, hasKey: status.llm.custom.hasKey)
                DesignDivider()
                SecretUrlRow(
                    hasValue: status.llm.custom.hasBaseUrl,
                    isEditing: editing == .customBaseUrl,
                    isSaving: isSaving,
                    onStartEdit: onStartEditBaseUrl,
                    onCancel: onCancelEdit,
                    onSave: onSaveBaseUrl
                )
            }
            if apply != .hidden {
                RestartNotice(phase: apply, onApply: onApplyNow, onDismiss: onDismissNotice)
            }
        } else if isError {
            AsyncNotice(kind: .error, title: "Couldn't load provider keys", detail: "Check your connection and try again.", retry: onRetryLoad)
        } else {
            AsyncNotice(kind: .loading, title: "Loading provider keys")
        }
    }

    private func providerSummary(_ status: SecretsStatus) -> some View {
        VStack(alignment: .leading, spacing: Space.sm) {
            Text("Selected provider · \(providerLabel(status.llm.active))")
                .designText(.label)
                .fontWeight(.semibold)
                .foregroundStyle(DuskColors.ink)
                .accessibilityAddTraits(.isHeader)
            Text("Provider selection is saved when selected; keys are saved with Save. Applying configuration does not test credentials or verify the running connection.")
                .designText(.supporting)
                .foregroundStyle(DuskColors.ink2)
        }
    }

    private func providerLabel(_ raw: String) -> String {
        switch SecretsViewModel.Provider(rawValue: raw) {
        case .openrouter: "OpenRouter"
        case .ollamaCloud: "Ollama Cloud"
        case .custom: "Custom"
        case nil: raw
        }
    }

    private func keyRow(
        _ provider: SecretsViewModel.Provider,
        label: String,
        status: SecretsStatus,
        hasKey: Bool
    ) -> some View {
        let active = status.llm.active == provider.rawValue
        return SecretKeyRow(
            label: label,
            idKey: provider.rawValue,
            hasKey: hasKey,
            isActive: active,
            isEditing: editing == .key(provider),
            isSaving: isSaving,
            onSetActive: active ? nil : { onSetActive(provider) },
            onStartEdit: { onStartEditKey(provider) },
            onCancel: onCancelEdit,
            onSave: { onSaveKey(provider, $0) }
        )
    }
}

private struct RestartNotice: View {
    let phase: SecretsViewModel.ApplyPhase
    let onApply: () -> Void
    let onDismiss: () -> Void

    var body: some View {
        VStack(alignment: .leading, spacing: Space.md) {
            AsyncNotice(kind: noticeKind, title: title, detail: detail)
            if phase != .applying && phase != .applied {
                HStack(spacing: Space.sm) {
                    DesignActionButton(
                        title: phase.isFailed ? "Retry apply" : "Apply now",
                        accessibilityId: "settings-secrets-apply",
                        action: onApply
                    )
                    DesignTextButton(
                        title: "Dismiss",
                        accessibilityId: "settings-secrets-notice-dismiss",
                        action: onDismiss
                    )
                }
            }
        }
        .accessibilityIdentifier("settings-secrets-restart-notice")
    }

    private var noticeKind: DesignNoticeKind {
        switch phase {
        case .applying: .loading
        case .applied: .success
        case .alreadyApplying: .warning
        case .failed: .error
        case .notice, .hidden: .warning
        }
    }
    private var title: String {
        switch phase {
        case .applying: "Applying configuration…"
        case .applied: "Configuration applied"
        case .alreadyApplying: "Already applying"
        case .failed: "Couldn't apply configuration"
        case .notice, .hidden: "Apply saved configuration"
        }
    }
    private var detail: String? {
        switch phase {
        case .alreadyApplying: "Try again when the current apply finishes."
        case .failed(let message): message
        case .notice: "Your provider selection and key changes are saved. Applying configuration does not test credentials or verify the running connection; saved credentials remain hidden."
        default: nil
        }
    }
}

extension SecretsViewModel.ApplyPhase {
    var isFailed: Bool { if case .failed = self { return true }; return false }
}

private func secretPreviewStatus(active: String) -> SecretsStatus {
    SecretsStatus(
        llm: LlmSecretsStatus(
            active: active,
            ollamaCloud: LlmProviderStatus(hasKey: true, hasBaseUrl: false),
            openrouter: LlmProviderStatus(hasKey: true, hasBaseUrl: false),
            custom: LlmProviderStatus(hasKey: false, hasBaseUrl: false)
        ),
        homeAssistant: HomeAssistantStatus(
            url: nil,
            observeToken: TokenPresence(hasToken: false),
            mcpServerToken: TokenPresence(hasToken: false)
        ),
        musicAssistant: MusicAssistantStatus(url: nil, hasToken: false)
    )
}

#Preview("Ready — large text") {
    NavigationStack {
        SecretsBody(
            status: secretPreviewStatus(active: "openrouter"), isError: false, isNotAdmin: false,
            mutationError: nil, editing: .none, isSaving: false, apply: .notice,
            onRetryLoad: {}, onSetActive: { _ in }, onStartEditKey: { _ in }, onStartEditBaseUrl: {},
            onCancelEdit: {}, onSaveKey: { _, _ in }, onSaveBaseUrl: { _ in }, onApplyNow: {}, onDismissNotice: {}
        )
    }
    .environment(\.dynamicTypeSize, .accessibility3)
}

#Preview("Not admin") {
    NavigationStack {
        SecretsBody(
            status: nil, isError: false, isNotAdmin: true, mutationError: nil, editing: .none,
            isSaving: false, apply: .hidden, onRetryLoad: {}, onSetActive: { _ in },
            onStartEditKey: { _ in }, onStartEditBaseUrl: {}, onCancelEdit: {},
            onSaveKey: { _, _ in }, onSaveBaseUrl: { _ in }, onApplyNow: {}, onDismissNotice: {}
        )
    }
}
